### PR TITLE
Single file held state

### PR DESCRIPTION
**Pull request by Haziba**
**Note to tester:**
Rejigged everything so state can be both held by the client and overwritten in each method call

### CODE REVIEW

**Comment by AdriVanHoudt** [comment]
This is a way nicer implementation :D 👌
* Would it make sense to make api like this: 
```js
const Cronofy = require('cronofy');
const cronofyClient = new Cronofy(options);
```
This would allow you to put the methods on the prototype which makes it even faster. In the cronofy function (https://github.com/cronofy/cronofy-node/pull/23/files#diff-1fdf421c05c1140f6d71444ea2b27638R13) you would have to do a `this.config = config` and where you use the config you'd have to use `this.config`
Does that make sense?
* It seems like it isn't possible to pass `next_page` to readEvents anymore (see https://github.com/cronofy/cronofy-node/pull/23/files#diff-a973e3c7958cb2dada1bee5d70a58e2eL12) which is something we definatly need

**Comment by Haziba** [comment]
Sound like some good changes to make, from looking at the `next_page` situation however I feel like the SDK would benefit from an iterator so that the pagination is handled from inside the Cronofy object?

Also I'm hesitant to make a breaking change to the error handling but currently we're passing back the whole response as the error, whereas I believe it would be better to just pass back the status code and the response body, so that whoever is using this client will just get the information they need and have an easier time working out what the issue is with a failing call.

What do you think to both these points?

**Comment by AdriVanHoudt** [comment]
* Handling next_page yourself seems fine. You'd need to keep track of it and provide a simple api to ask for the next page then.
* Decent errors would be great! I now need to `delete error.raw` to get some valuable errors :P

File: src/index.js [hunk 1]
```diff
@@ -11,6 +11,8 @@ var tap = function(func){
 }
 
 var cronofy = function(config){
+  this.config = config;
+
   this.accountInformation = function(){
```
Comment (AdriVanHoudt) [inline src/index.js]: this sets it on the prototype due to using new? TIL :D
You might want to assert here that the function is being called using new

File: src/index.js [hunk 2]
```diff
@@ -176,10 +176,14 @@ var cronofy = function(config){
           resolve(result['entity']);
         }
       }, function(err){
+        var error = {
```
Comment (AdriVanHoudt) [inline src/index.js]: I would prefer a real Error object here.
Is err.entity a real error? if so just add the status code to that and give that back

**Comment by Haziba** [comment]
@AdriVanHoudt How's that?

File: src/index.js [hunk 3]
```diff
@@ -11,6 +11,11 @@ var tap = function(func){
 }
 
 var cronofy = function(config){
+  if(!this || this.constructor !== cronofy){
+    console.log("Cronofy must be called as a constructor");
```
Comment (AdriVanHoudt) [inline src/index.js]: Why not throw here? If you are not doing new you are doing something wrong no?

**Comment by AdriVanHoudt** [comment]
Are you sure that the current implementation does the prototype thing?
I think you are still setting all the methods on creating a new instance. Instead of reusing them for each instance.
What you want to do (I think) is
```js
var cronofy = function cronofy(config) {
     this.config = config;
};

cronofy.prototype.readEvents = function (...) {...};
```
Does that make sense or am I missing the ball here?

**Comment by Haziba** [comment]
Ah sorry I misunderstood! JavaScript isn't my primary language so I'm not quite so knowledgable on prototypes and the like, so thanks very much for your help :)

Does this look alright now?

File: src/index.js [hunk 4]
```diff
@@ -17,147 +17,19 @@ var cronofy = function(config){
...
-      delete config.refresh_token;
-
-      if(details.callback){
-        details.callback();
-      }
-    }), details.callback);
-  }
-
   var urls = {
     api: 'http://local' + (config.dataCenter ? '-' + config.dataCenter : '') + '.cronofy.com'
```
Comment (AdriVanHoudt) [inline src/index.js]: If you put this on `this` you can move the `httpGet/Post/Delete` methods to the prototype as well.
Also I would prefix them with `_` to let people know they're not public methods.

If you can pass the `urls.api` to the `httpCall` function you can even move that off the prototype to just be an internal function. (my own personal rule is to put as little as needed on the "public" object)

**Comment by Haziba** [comment]
@AdriVanHoudt How's that?

**Comment by AdriVanHoudt** [comment]
Seems good 👍

**Pull request merged**